### PR TITLE
cli: Improve pruning documentation

### LINKED ISSUE #12811: Confusing blocks-pruning description
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

>  Default is to keep all finalized blocks. otherwise, all blocks can be kept (i.e 'archive'), or for all canonical blocks (i.e 'archive-canonical'), or for the last N blocks (i.e a number).
> 
> NOTE: only finalized blocks are subject for removal!

If only the finalized blocks are subject to removal, and all finalized blocks are canonical (maybe I'm wrong here?), what is the difference between parameter default (all finalized blocks) and `archive-canonical` ?

### Steps to reproduce

_No response_

### PR DESCRIPTION
**Pull request by lexnv**
This PR improves the documentation of the pruning modes
exposed as CLI flags.

This specifies what the options mean (ie archive, archive-canonical, or number),
and what the two pruning modes are responsible for (state vs block pruning).

While at it, remove the old note that stated that only finalized blocks
are subject to pruning (which is not entirely true anymore, since we are
also pruning displaced branches -- forks).

Closes #12811.

### CODE REVIEW

File: client/cli/src/params/pruning_params.rs [hunk 1]
```diff
@@ -23,21 +23,30 @@ use sc_service::{BlocksPruning, PruningMode};
...
+	/// should be pruned (ie, removed) from the database.
 	///
-	/// Default is to keep all finalized blocks.
-	/// otherwise, all blocks can be kept (i.e 'archive'),
-	/// or for all canonical blocks (i.e 'archive-canonical'),
-	/// or for the last N blocks (i.e a number).
+	/// Options available:
+	///   'archive'            Keep all blocks.
+	///   'archive-canonical'  Keep only finalized blocks.
+	///   [number]             Keep the last number of finalized blocks.
```
Comment (lexnv) [inline client/cli/src/params/pruning_params.rs:47]:
```suggestion
	///   number               Keep the last number of finalized blocks.
```

File: client/cli/src/params/pruning_params.rs [hunk 2]
```diff
@@ -23,21 +23,30 @@ use sc_service::{BlocksPruning, PruningMode};
...
+	/// should be pruned (ie, removed) from the database.
+	///
+	/// Options available:
+	///   'archive'            Keep the state of all blocks.
+	///   'archive-canonical'  Keep only the state of finalized blocks.
+	///   number               Keep the state of the last number of finalized blocks.
+	///
+	/// The default option is to keep the last 256 blocks (number == 256).
 	#[arg(alias = "pruning", long, value_name = "PRUNING_MODE")]
 	pub state_pruning: Option<String>,
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:38]: We should change this to:
```suggestion
	#[arg(alias = "pruning", long, value_name = "PRUNING_MODE", default = "256")]
	pub state_pruning: PruningMode,
```

Then add the following enum:

```
enum PruningMode {
     Archive,
     ArchiveCanonical,
     Custom(u32),
}

impl clap::ValueEnum for PruningMode {
     .. impl the enum
}
```

File: client/cli/src/params/pruning_params.rs [hunk 3]
```diff
@@ -23,21 +23,30 @@ use sc_service::{BlocksPruning, PruningMode};
...
-	/// or for the last N blocks (i.e a number).
+	/// Options available:
+	///   'archive'            Keep all blocks.
+	///   'archive-canonical'  Keep only finalized blocks.
+	///   number               Keep the last number of finalized blocks.
 	///
-	/// NOTE: only finalized blocks are subject for removal!
+	/// The default option is 'archive-canonical'.
 	#[arg(alias = "keep-blocks", long, value_name = "COUNT")]
 	pub blocks_pruning: Option<String>,
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:51]: We can then do the same here.

File: client/cli/src/params/pruning_params.rs (hunk 2 above)
Comment (lexnv) [reply client/cli/src/params/pruning_params.rs]: Yep, that does make sense! Thanks for the suggestion!

One problem with `ValueEnum` is that its happy with unit only enum variants. This is related to the [value_variants](https://docs.rs/clap/latest/clap/trait.ValueEnum.html#tymethod.value_variants) that requires us to state the expected values. If the `Custom(32)` is omitted, the default value of "256" cannot be provided anymore.

To overcome this I've changed the `ValueEnum` to `FromStr`.
Let me know what you think of it

File: client/cli/src/params/pruning_params.rs [hunk 4]
```diff
@@ -23,57 +23,78 @@ use sc_service::{BlocksPruning, PruningMode};
...
-	/// otherwise, all blocks can be kept (i.e 'archive'),
-	/// or for all canonical blocks (i.e 'archive-canonical'),
-	/// or for the last N blocks (i.e a number).
-	///
-	/// NOTE: only finalized blocks are subject for removal!
-	#[arg(alias = "keep-blocks", long, value_name = "COUNT")]
-	pub blocks_pruning: Option<String>,
+	/// This mode specifies when the block's body (including justifications)
+	/// should be pruned (ie, removed) from the database.
+	#[arg(alias = "keep-blocks", long, value_name = "COUNT", default_value = "archive-canonical")]
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:36]:
```suggestion
	#[arg(alias = "keep-blocks", long, value_name = "PRUNING_MODE", default_value = "archive-canonical")]
```

File: client/cli/src/params/pruning_params.rs [hunk 5]
```diff
@@ -23,57 +23,78 @@ use sc_service::{BlocksPruning, PruningMode};
...
-	/// Default is to keep only the last 256 blocks,
-	/// otherwise, the state can be kept for all of the blocks (i.e 'archive'),
-	/// or for all of the canonical blocks (i.e 'archive-canonical').
-	#[arg(alias = "pruning", long, value_name = "PRUNING_MODE")]
-	pub state_pruning: Option<String>,
-	/// Specify the blocks pruning mode, a number of blocks to keep or 'archive'.
+	/// This mode specifies when the block's state (ie, storage)
+	/// should be pruned (ie, removed) from the database.
+	#[arg(alias = "pruning", long, value_name = "PRUNING_MODE", default_value = "256")]
+	pub state_pruning: DatabasePruningMode,
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:31]: We need to tell clap that `DatabasePruningMode` is an enum value. So, that it shows the possible values you can pass to it. As I said in my initial comment, you probably need to implement this trait as the derive macro will maybe not deliver good results. However, you can first try the macro and look at the results :)

**Review by bkchr — COMMENTED** [review]
Already looks much better, but now we just need to tell `clap` that these values are enum_values.

File: client/cli/src/params/pruning_params.rs [hunk 6]
```diff
@@ -23,57 +23,97 @@ use sc_service::{BlocksPruning, PruningMode};
...
 	///
-	/// Default is to keep all finalized blocks.
-	/// otherwise, all blocks can be kept (i.e 'archive'),
-	/// or for all canonical blocks (i.e 'archive-canonical'),
-	/// or for the last N blocks (i.e a number).
+	/// Possible values:
+	///  'archive'            Keep the state of all blocks.
+	///  'archive-canonical'  Keep only the state of finalized blocks.
+	///   number              Keep the state of the last number of finalized blocks.
 	///
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:35]:
```suggestion
```

File: client/cli/src/params/pruning_params.rs [hunk 7]
```diff
@@ -23,57 +23,97 @@ use sc_service::{BlocksPruning, PruningMode};
...
-	/// NOTE: only finalized blocks are subject for removal!
-	#[arg(alias = "keep-blocks", long, value_name = "COUNT")]
-	pub blocks_pruning: Option<String>,
+	/// The default option is to keep the last 256 blocks (number == 256).
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:36]:
```suggestion
```

This is already done by `default_value = 256`

File: client/cli/src/params/pruning_params.rs [hunk 8]
```diff
@@ -23,57 +23,97 @@ use sc_service::{BlocksPruning, PruningMode};
...
+	///
+	/// This mode specifies when the block's body (including justifications)
+	/// should be pruned (ie, removed) from the database.
+	///
+	/// Possible values:
+	///  'archive'            Keep all blocks.
+	///  'archive-canonical'  Keep only finalized blocks.
+	///   number              Keep the last number of finalized blocks.
+	///
+	/// The default option is 'archive-canonical'.
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:49]:
```suggestion
```

File: client/cli/src/params/pruning_params.rs [hunk 9]
```diff
@@ -23,57 +23,97 @@ use sc_service::{BlocksPruning, PruningMode};
...
+	pub state_pruning: DatabasePruningMode,
+	/// Specify the blocks pruning mode.
+	///
+	/// This mode specifies when the block's body (including justifications)
+	/// should be pruned (ie, removed) from the database.
+	///
+	/// Possible values:
+	///  'archive'            Keep all blocks.
+	///  'archive-canonical'  Keep only finalized blocks.
+	///   number              Keep the last number of finalized blocks.
```
Comment (bkchr) [inline client/cli/src/params/pruning_params.rs:47]:
```suggestion
	///   number              Keep the last `number` of finalized blocks.
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
Thank you @lexnv

**Pull request merged**